### PR TITLE
feat(express): auto-track errors

### LINKED ISSUE #84: Automate adding of Express agent middleware
If you're running express with Elastic APM, you today need to add our [custom error catching middleware](https://www.elastic.co/guide/en/apm/agent/nodejs/current/express.html#express-error-handlers) manually.

This have confused quite a few people and is unique to how Express compare to our setup instructions for instrumenting other Node.js frameworks, so it would be nice if middleware was added automatically when starting the agent, similar to how Express is patched in other ways.

### PR DESCRIPTION
**Pull request by Qard**
This enables automatic tracking of express errors, rather than requiring manual use of the middleware.

Fixes #84

### CODE REVIEW

File: lib/instrumentation/modules/express.js [hunk 1]
```diff
@@ -12,12 +12,45 @@ module.exports = function (express, agent, version) {
...
+  function patchLayer (layer) {
+    if (layerPatched) return
+    layerPatched = true
+
+    debug('shimming express.Router.Layer.handle_error function')
+
+    shimmer.wrap(layer.constructor.prototype, 'handle_error', function (orig) {
+      return function (err, req) {
+        if (!err[reportedSymbol]) {
+          err[reportedSymbol] = true
```
Comment (watson) [inline lib/instrumentation/modules/express.js]: I really like this approach. I can think of a few places in our code base where we should do that instead 😅

File: lib/instrumentation/modules/express.js [hunk 2]
```diff
@@ -12,12 +12,45 @@ module.exports = function (express, agent, version) {
     return express
   }
 
+  // express 5 moves the router methods onto a prototype
```
Comment (watson) [inline lib/instrumentation/modules/express.js]: Nice that you're thinking of Express 5 already 👍

Comment (Qard) [reply lib/instrumentation/modules/express.js]: I haven't fully tested express 5, but that was an easy part to fix.

**Comment by watson** [comment]
We should probably also add the new Express test file file to .tav.yml 😃

**Comment by Qard** [comment]
Since it's in end-user code, would it be better to have a deprecation warning for a bit?

**Comment by watson** [comment]
Normally I would deprecate yes, but as we're still in beta and since we're making a few other breaking changes when go GA, I think it's ok for us to make this one breaking as well. The more we can clean up now and not carry around as legacy for a while the cleaner/better I think. We'll have to make an upgrade guide for both old Opbeat users and for beta users when we go into GA anyway, so we can have a section mentioning breaking changes like this here. That's at least my current philosophy 😄

**Comment by Qard** [comment]
Made those changes and also, when adding the TAV details discovered the approach I took didn't work before express 4.6 so I made some changes to how the patch gets applied. It's not using the `layer.constructor.prototype` anymore, so each layer instance will get patched separately, but as far as I can tell that should only happen the first time that layer is used.

File: docs/agent-api.asciidoc [hunk 3]
```diff
@@ -642,11 +642,9 @@ In which case it will automate this processes for you.
 apm.middleware.connect()
 ----
 
-Alias: `apm.middleware.express()`
-
 Returns a middleware function used to collect and send errors to the APM Server.
 
-The middleware can be used as-is with either Connect or Express in the same way.
+The middleware can be used as-is with either Connect in the same way.
```
Comment (watson) [inline docs/agent-api.asciidoc]: I'd just remove the entire line here

**Comment by watson** [comment]
@Qard if you feel it's better I think we should consider just dropping support for Express <4.6 as that version was released in June 2014. What do you think?

File: docs/agent-api.asciidoc (hunk 3 above)
Comment (Qard) [reply docs/agent-api.asciidoc]: fixed.

**Comment by watson** [comment]
@Qard As you also mentioned in https://github.com/elastic/apm-agent-nodejs/pull/131#issuecomment-352911890, we forgot to add the new tav test to the `TAV` environment variables. That should be the last thing missing 😃

Before merging we need to figure out if we want to publish a new version of the module right after (as it's changing the docs), or if there's a better way to handle this. See #134 for more details.

Tip: I just tested it locally on Node 9 by having `the-all-versions` module installed globally and then running this command:

```
tav express '>=4' -- node test/instrumentation/modules/express.js
```

... and it all worked perfectly 👍

**Comment by watson** [comment]
Weird... I tried to resolve the conflicts using the "Resolve conflicts" button, but even though the commit that resolves them are included in the PR, GitHub still thinks the conflicts exists.

**Pull request merged**